functorch/functorch/csrc/dim/arena.h

```diff
@@ -0,0 +1,328 @@
+        }
+        return idx.has_value();
+    }
+
+    Slice slice(int begin) {
+        return slice(begin, size_);
+    }
+
+    Slice slice(int begin, int end) {
+        if (begin < 0) {
+            begin += size_;
+        }
+        if (end < 0) {
+            end += size_;
+        }
+        Slice result;
+        result.begin_ = begin_ + begin;
+        result.size_ = end - begin;
+        result.capacity_ = result.size_;
+        return result;
+    }
+
+    bool inside(Slice where) {
+        return begin() <= where.begin() && where.end() <= end();
+    }
+
+    irange enumerate() const {
+        return irange(size_);
+    }
+
+    irange reversed_enumerate() const {
+        return irange(size_ - 1, -1, -1);
+    }
+
+    bool operator==(const Slice<T>& rhs) const {
+        if (size() != rhs.size()) {
+            return false;
+        }
+        return std::equal(begin(), end(), rhs.begin());
+    }
+
+    Slice(T* begin, T* end)
+    : begin_(begin), size_(end - begin), capacity_(size_) {}
+
+protected:
+    static int _length(const T& t) {
+        return 1;
+    }
+    static int _length(Slice t) {
+        return t.size_;
+    }
+    static T* _insert(T*& dst, T t) {
+        *dst = std::move(t);
+        return ++dst;
+    }
+    static T* _insert(T*& dst, Slice t) {
+        std::memcpy(dst, t.begin_, sizeof(T)*t.size_);
+        dst += t.size_;
+        return dst;
+    }
+    T* begin_;
+    int size_;
+    int capacity_;
+    friend struct OwnedSlice<T>;
+};
+
+template<typename T>
+struct OwnedSlice {
+    typedef void (*deleter_t)(Slice<T>);
+    static void _no_delete(Slice<T>) {}
+    OwnedSlice()
+    : deleter_(_no_delete) {}
+    OwnedSlice(const OwnedSlice&) = delete;
+    OwnedSlice& operator=(const OwnedSlice&) = delete;
+    ~OwnedSlice() {
+        deleter_(slice_);
+        if (slice_.size_ > 8) {
+            delete [] slice_.begin_;
+        }
+    }
+    void set(Slice<T> to_own, deleter_t deleter = _no_delete) {
+        slice_.size_ = slice_.capacity_ = to_own.size();
+        slice_.begin_ = (slice_.size_ > 8) ? new T[slice_.size_] : &small_buf[0];
+        std::memcpy(slice_.begin_, to_own.begin(), slice_.size_ * sizeof(T));
+        deleter_ = deleter;
+    }
+    Slice<T> slice() const {
+        return slice_;
+    }
+private:
+    Slice<T> slice_;
+    deleter_t deleter_;
+    T small_buf[8];
+};
+
+template<typename T>
+inline std::ostream& operator<<(std::ostream& s, const Slice<T>& v) {
+    s << "[";
+    for (int i : v.enumerate()) {
+        if (i > 0) {
+            s << ", ";
+        }
+        s << v[i];
+    }
+    s << "]";
+    return s;
+}
+
+struct TensorRef {
+    TensorRef()
+    : impl_(nullptr){}
+    TensorRef(const at::Tensor& t)
+    : impl_(t.unsafeGetTensorImpl()) {}
+    const at::Tensor& operator*() const {
+        return *(at::Tensor*)this;
+    }
+    at::Tensor* operator->() const {
+        return (at::Tensor*)this;
+    }
```

Comment:
these casts have undefined behavior, and because of TBAA I wouldn't be surprised at all if you were actually bitten by this in practice.

see the aliasing rules at https://en.cppreference.com/w/cpp/language/reinterpret_cast and note that C-style casts are interpreted as whichever fancy_cast matches, which will be `reinterpret_cast` in this case https://en.cppreference.com/w/cpp/language/explicit_cast